In [2]:
all_ops = {
    "+1": (lambda x: x + 1, "-1"),
    "-1": (lambda x: x - 1, "+1"),
    "*2": (lambda x: x * 2, "/2"),
    "/2": (lambda x: x / 2, "*2"),
    "exp": (lambda x: exp(x), "log"),
    "log": (lambda x: log(x) if x > 0 else None, "exp"),
    "*pi": (lambda x: x * RR(pi), "/pi"),
    "/pi": (lambda x: x * RR(pi), "*pi"),
    "1/": (lambda x: 1/x if x != 0 else None, "1/"),
    "sqrt": (lambda x: sqrt(x), "square"),
    "square": (lambda x: x^2, "sqrt"),
    "negate": (lambda x: -x, "negate")
}

In [3]:
search_ops = {
    #"+1": (lambda x: x + 1, "-1"),
    #"-1": (lambda x: x - 1, "+1"),
    "*2": (lambda x: x * 2, "/2"),
    "/2": (lambda x: x / 2, "*2"),
    "exp": (lambda x: exp(x), "log"),
    "log": (lambda x: log(x) if x > 0 else None, "exp"),
    "*pi": (lambda x: x * RR(pi), "/pi"),
    #"/pi": (lambda x: x * RR(pi), "*pi"),
    "1/": (lambda x: 1/x if x != 0 else None, "1/"),
    "sqrt": (lambda x: sqrt(x), "square"),
    "square": (lambda x: x^2, "sqrt"),
    "negate": (lambda x: -x, "negate")
}

In [4]:
class NumContainer():
    
    def __init__(self, n, hist, prec):
        self.n = n
        self.hist = hist
        self.prec = prec
    
    def __hash__(self):
        return hash(round(self.n / (2*self.prec))*2*self.prec)
    
    def __eq__(self, other):
        return round(self.n / self.prec) == round(other.n / self.prec)
    
    def __str__(self):
        return str(self.n) + " with history " + str(self.hist)
    


def search(numbers, prec = 0.00001, depth = 3):
    sets = [{NumContainer(n, [n], prec): NumContainer(n, [n], prec)} for n in numbers]
    
    for d in range(depth):
        for i in range(len(numbers)):
            additions = []
            for n0 in sets[i]:
                for name in search_ops:
                    op = search_ops[name][0]
                    nval = op(n0.n)
                    if nval in RR and not (nval == None or RR(nval).is_NaN() or RR(nval).is_infinity()):
                        if log(abs(nval)) < 1000 and log(abs(nval)) > -10:
                            n1 = NumContainer(nval, n0.hist + [name],n0.prec)
                            additions.append(n1)
                            for i1 in range(len(numbers)):
                                if i1 != i:
                                    if n1 in sets[i1]:
                                        n2 = sets[i1][n1]
                                        yield (n1, n2)
            for add in additions:
                 sets[i][add] = add
    

def search_print(numbers, prec = 0.00001, depth = 3):
    l = []
    for n1, n2 in search(numbers, prec, depth):
        print(n1, "is connected to ", n2)
        l.append((n1, n2))
    return l

In [5]:
def isolate(n1, n2):
    lefthist = n1.hist[:0:-1]
    newhist = [all_ops[act][1] for act in lefthist]
    return n2.hist + newhist

In [10]:
search_print([RR(389), RR(7.96892424103156)], prec = 10^(-10),depth = 5)

[]

In [8]:
search_print([RR(11), RR(28.9591226820056)], prec = 10^(-10),depth = 6)

[]

log(-1/x) = pi*sqrt(1/289)
x = -exp(-pi/sqrt(289))

In [18]:
search_print([RR(11), RR(0.24027768804459423)], prec = 10^(-7),depth = 5)

0.000134381940515793 with history [0.240277688044594, '+1', 'log', '/2', 'square', 'square'] is connected to  0.000134372480515990 with history [11.0000000000000, 'square', '+1', 'square', '/2', '1/']
0.0289371898709240 with history [0.240277688044594, '/2', 'exp', '*pi', 'exp', '1/'] is connected to  0.0289372184442784 with history [11.0000000000000, 'exp', '+1', 'log', '*pi', '1/']
0.0898513522555034 with history [0.240277688044594, '/2', 'exp', '*pi', '*pi', '1/'] is connected to  0.0898514158086345 with history [11.0000000000000, '*pi', 'log', '*pi', '1/']
0.0289371898709240 with history [0.240277688044594, '/2', 'exp', '*pi', 'negate', 'exp'] is connected to  0.0289372184442784 with history [11.0000000000000, 'exp', '+1', 'log', '*pi', '1/']
0.0398399683655546 with history [0.240277688044594, 'exp', '*2', '*pi', '*pi', '1/'] is connected to  0.0398400247243197 with history [11.0000000000000, '*pi', 'log', 'square', '*2', '1/']
0.0253629115920110 with history [0.240277688044594, 'e

[(<__main__.NumContainer object at 0x7fed45105be0>,
  <__main__.NumContainer object at 0x7fed449cff98>),
 (<__main__.NumContainer object at 0x7fed4468e048>,
  <__main__.NumContainer object at 0x7fed45b7c0f0>),
 (<__main__.NumContainer object at 0x7fed4468e4a8>,
  <__main__.NumContainer object at 0x7fed46012320>),
 (<__main__.NumContainer object at 0x7fed4468ea90>,
  <__main__.NumContainer object at 0x7fed45b7c0f0>),
 (<__main__.NumContainer object at 0x7fed444abdd8>,
  <__main__.NumContainer object at 0x7fed447f2828>),
 (<__main__.NumContainer object at 0x7fed45b13c18>,
  <__main__.NumContainer object at 0x7fed447f2e48>),
 (<__main__.NumContainer object at 0x7fed45b13d68>,
  <__main__.NumContainer object at 0x7fed447f2828>),
 (<__main__.NumContainer object at 0x7fed459d1278>,
  <__main__.NumContainer object at 0x7fed444dae80>)]

0.0898513522555034 with history (0.240277688044594, '/2', 'exp', '*pi', '*pi', '1/') is connected to  0.0898514158086345 with history (11.0000000000000, '*pi', 'log', '*pi', '1/')

1/(exp(x/2) * pi * pi) = 1/(log(11*pi)*pi)
x = 2log(log(11*pi)/pi)

In [20]:
RR(2*log(log(11*pi)/pi))

0.240276273416834

In [7]:
search_print([RR(389), RR(e^(-2*pi/sqrt(389)))], prec = 0.000001)

0.0253717336788930 with history [0.727188070492475, 'log', '/2', 'square'] is connected to  0.0253717336788929 with history [389.000000000000, '1/', '*pi', '*pi']
0.159285070483373 with history [0.727188070492475, 'log', '/2', 'negate'] is connected to  0.159285070483373 with history [389.000000000000, '1/', 'sqrt', '*pi']
-0.000817614514191156 with history [0.727188070492475, 'log', '*pi', '+1'] is connected to  -0.000818277342374783 with history [389.000000000000, '*pi', 'negate', '1/']
1.00081761451419 with history [0.727188070492475, 'log', '*pi', 'negate'] is connected to  1.00081827734237 with history [389.000000000000, '*pi', '1/', '+1']
-9.86154146165801 with history [0.727188070492475, 'log', '1/', '*pi'] is connected to  -9.86154146165801 with history [389.000000000000, 'sqrt', 'negate', '/2']
0.159285070483373 with history [0.727188070492475, 'log', 'negate', '/2'] is connected to  0.159285070483373 with history [389.000000000000, '1/', 'sqrt', '*pi']
1.00081761451419 with h

[(<__main__.NumContainer object at 0x7fed46126f98>,
  <__main__.NumContainer object at 0x7fed4617e940>),
 (<__main__.NumContainer object at 0x7fed46126fd0>,
  <__main__.NumContainer object at 0x7fed4617eb70>),
 (<__main__.NumContainer object at 0x7fed461cc080>,
  <__main__.NumContainer object at 0x7fed4617e2e8>),
 (<__main__.NumContainer object at 0x7fed461cc1d0>,
  <__main__.NumContainer object at 0x7fed46179c18>),
 (<__main__.NumContainer object at 0x7fed461cc320>,
  <__main__.NumContainer object at 0x7fed46185b70>),
 (<__main__.NumContainer object at 0x7fed461cc668>,
  <__main__.NumContainer object at 0x7fed4617eb70>),
 (<__main__.NumContainer object at 0x7fed461cc710>,
  <__main__.NumContainer object at 0x7fed46179c18>),
 (<__main__.NumContainer object at 0x7fed4613a128>,
  <__main__.NumContainer object at 0x7fed4617eb70>)]

In [43]:
#search_print([RR(389), RR(exp(-2*pi/sqrt(389)))], prec = 10^(-10));

In [105]:
#res = list(search([RR(389), RR(-0.6042576046984)], prec = 10^(-10), depth = 7))
res = list(search([RR(389), RR(0.385172566491)], prec = 10^(-8), depth = 5))

In [108]:
for n1, n2 in res:
    if n1.hist[0] == 389:
        hist = isolate(n2, n1)
    else:
        hist = isolate(n1, n2)
    
    #n = hist[0]
    
    #for action in hist[1:]:
        #op = all_ops[action][0]
        #n = op(n)
        
    diff = log(abs(n1.n - n2.n))
    
    print(diff, hist)
    
    #print(n1, "is connected to ", n2)

-19.5670082223284 [389.000000000000, 'log', '+1', '+1', 'log', 'sqrt', 'exp', '-1', '/2', 'square', '1/']


1-3 are the same
4 is different

1:
$$\exp(x * \pi * \pi) = --(1/389)$$
$$\exp(x * \pi * \pi) = 1/389$$
$$x = \log(1/389)/\pi^2$$

2:
$$1/(x * \pi * \pi) = -1/\log(389)$$
$$x = -\log(389)/\pi^2$$

3:
$$(exp(x * pi * pi)*pi)^2 = --(pi/389)^2$$
$$exp(x * pi * pi)*pi = (+-)pi/389$$
$$x * pi * pi = log((+-)1/389)$$
$$x = log((+-)1/389)/pi^2$$

4:
$$((x/2/2)^2/2)^2 = (1/(pi*sqrt(389*2)))^2$$
$$(x/2/2)^2/2 = (+-)1/(pi*sqrt(389*2))$$
$$x = (+-)4sqrt(2/(pi*sqrt(389*2)))$$

5:
0.604257376655331 (389.000000000000, 'log', '1/', 'exp', 'sqrt', 'square', '-1', '*2', 'sqrt')
$$sqrt((sqrt(exp(1/log(389)))^2 - 1) * 2)$$
$$sqrt((exp(1/log(389)) - 1) * 2)$$

6:
-0.604257820290370 (389.000000000000, '/2', '+1', '*pi', '1/', '*2', 'exp', 'exp', '-1', 'log', 'log')
$$log(log(exp(exp(2/(pi*(389/2 + 1)))) - 1))$$

7:
-0.604257652448222 (389.000000000000, '+1', 'exp', '/2', '/2', 'log', '1/', '/2', 'sqrt', 'sqrt', 'exp', 'negate', '/2')
-exp(sqrt(sqrt((1/log(exp(389 + 1)/2/2))/2))))/2
-exp(sqrt(sqrt((1/log(exp(389 + 1)/4))/2))))/2

8:
-0.604257604164496 (389.000000000000, '1/', 'sqrt', '*2', 'exp', '1/', '*pi', '-1', '-1', 'exp', '/2', '-1', 'sqrt', '-1')
sqrt(exp(pi/exp(sqrt(1/389)*2) - 1 - 1)/2 - 1) - 1

9:
1.89833121779761 (389.000000000000, 'sqrt', '/2', '*pi', '*pi', '+1', 'square', '1/', 'sqrt', '1/', 'log', '/pi', 'sqrt', '/2')
sqrt(log(1/(sqrt(1/square(pi*pi*sqrt(389)/2 + 1))))/pi)/2

10:
(389.000000000000, 'log', '+1', '+1', 'log', 'sqrt', 'exp', '-1', '/2', 'square', '1/')
1/((exp(sqrt(log(log(389) + 2))) - 1)/2)^2

In [ ]:
# Actual value:
-0.6042576046984
0.385172566491

In [15]:
RR(-log(389)/(pi^2))
# Not good

-0.604236917840417

In [41]:
RR(-4*sqrt(2/(pi*sqrt(389*2))))
# Even worse

-0.604303561177701

In [77]:
RR(-sqrt(2*(exp(1/log(389)) - 1)))
# Pretty good

0.604257376655331

In [84]:
RR(log(log(exp(exp(2*1/(pi*(389/2 + 1)))) - 1)))

-0.604257820290370

In [91]:
RR(-exp(sqrt(sqrt((1/log(exp(389 + 1)/4))/2)))/2)


-0.604257652448222

In [101]:
RR(sqrt(exp(pi/exp(sqrt(1/389)*2) - 1 - 1)/2 - 1) - 1)

-0.604257604164496

In [98]:
RR(-sqrt(log(1/(sqrt(1/(pi*pi*sqrt(389)/2 + 1)^2)))/pi)/2)

-0.604257593876293

In [109]:
RR(1/((exp(sqrt(log(log(389) + 2))) - 1)/2)^2)

0.385172569698865

In [13]:
def searchForInteger(n, prec = 0.00001, depth = 3):
    
    ops = [
        #(lambda x: x + 1, "+1"),
        #(lambda x: x - 1, "-1"),
        (lambda x: x * 2, "*2"),
        (lambda x: x / 2, "/2"),
        (lambda x: exp(x), "exp"),
        (lambda x: log(x) if x > 0 else None, "log"),
        (lambda x: x * RR(pi), "*pi"),
        (lambda x: 1/x if x != 0 else None, "1/"),
        (lambda x: sqrt(x), "sqrt"),
        (lambda x: x^2, "square"),
        (lambda x: -x, "negate")
    ]
    
    numbers = [(n, [n])]
    for d in range(depth):
        additions = []
        
        for n1, hist in numbers:
            for op, name in ops:
                n2 = op(n1)
                if n2 in RR and not (RR(n2).is_NaN() or RR(n2).is_infinity()):
                    if log(abs(n2)) < 300:
                        nhist = hist + [name]
                        additions.append((n2, nhist))
                        if (abs(n2 - round(n2)) < prec):
                            yield (n2, nhist)
        
        numbers.extend(additions)
        
        
def searchForInteger_print(n, prec = 0.00001, depth = 3):
    l = []
    for n1, hist in searchForInteger(n, prec, depth):
        print(n1, "with history", hist)
        l.append((n1, hist))
    return l

In [17]:
searchForInteger_print(-0.6042576046984, depth = 5)

6.60577012543628e-6 with history [-0.604257604698400, '*2', '*pi', '*pi', 'exp']
6.60577012543628e-6 with history [-0.604257604698400, '*pi', '*2', '*pi', 'exp']
6.60577012543628e-6 with history [-0.604257604698400, '*pi', '*pi', '*2', 'exp']
6.60577012543628e-6 with history [-0.604257604698400, '*pi', '*pi', 'exp', 'square']
7.29717552932348e-9 with history [-0.604257604698400, '*pi', '*pi', '*pi', 'exp']
8.96842108287975e15 with history [-0.604257604698400, '*pi', 'square', 'exp', 'exp']
8.06264996849856e-8 with history [-0.604257604698400, '1/', '*pi', '*pi', 'exp']
6.60577012543628e-6 with history [-0.604257604698400, '*2', '*pi', '*pi', 'exp']
6.60577012543628e-6 with history [-0.604257604698400, '*pi', '*2', '*pi', 'exp']
6.60577012543628e-6 with history [-0.604257604698400, '*pi', '*pi', '*2', 'exp']
6.60577012543628e-6 with history [-0.604257604698400, '*pi', '*pi', 'exp', 'square']
7.29717552932348e-9 with history [-0.604257604698400, '*pi', '*pi', '*pi', 'exp']
8.968421082879

8.09878036966054e103 with history [-0.604257604698400, 'square', '/2', '1/', 'exp', 'exp']
1.23399679404159e-6 with history [-0.604257604698400, 'square', '/2', 'square', 'square', 'square']
4.51142790806771e29 with history [-0.604257604698400, 'square', 'exp', 'exp', 'exp', 'exp']
1.33995845431078e40 with history [-0.604257604698400, 'square', 'exp', '*pi', 'exp', 'exp']
8.96842108287975e15 with history [-0.604257604698400, 'square', '*pi', '*pi', 'exp', 'exp']
8.09878036966054e103 with history [-0.604257604698400, 'square', '1/', '*2', 'exp', 'exp']
1.91578883268845e-7 with history [-0.604257604698400, 'square', '1/', 'exp', 'exp', '1/']
1.27094662370924e21 with history [-0.604257604698400, 'square', '1/', 'exp', '*pi', 'exp']
8.09878036966077e103 with history [-0.604257604698400, 'square', '1/', 'exp', 'square', 'exp']
1.91578883268845e-7 with history [-0.604257604698400, 'square', '1/', 'exp', 'negate', 'exp']
1.41605253797069e32 with history [-0.604257604698400, 'square', '1/', '*

[(6.60577012543628e-6, [-0.604257604698400, '*2', '*pi', '*pi', 'exp']),
 (6.60577012543628e-6, [-0.604257604698400, '*pi', '*2', '*pi', 'exp']),
 (6.60577012543628e-6, [-0.604257604698400, '*pi', '*pi', '*2', 'exp']),
 (6.60577012543628e-6, [-0.604257604698400, '*pi', '*pi', 'exp', 'square']),
 (7.29717552932348e-9, [-0.604257604698400, '*pi', '*pi', '*pi', 'exp']),
 (8.96842108287975e15, [-0.604257604698400, '*pi', 'square', 'exp', 'exp']),
 (8.06264996849856e-8, [-0.604257604698400, '1/', '*pi', '*pi', 'exp']),
 (6.60577012543628e-6, [-0.604257604698400, '*2', '*pi', '*pi', 'exp']),
 (6.60577012543628e-6, [-0.604257604698400, '*pi', '*2', '*pi', 'exp']),
 (6.60577012543628e-6, [-0.604257604698400, '*pi', '*pi', '*2', 'exp']),
 (6.60577012543628e-6, [-0.604257604698400, '*pi', '*pi', 'exp', 'square']),
 (7.29717552932348e-9, [-0.604257604698400, '*pi', '*pi', '*pi', 'exp']),
 (8.96842108287975e15, [-0.604257604698400, '*pi', 'square', 'exp', 'exp']),
 (8.06264996849856e-8, [-0.604257